clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java

```diff
@@ -52,36 +50,21 @@ public void configure(Map<String, ?> configs) {}
      * @param cluster The current cluster metadata
      */
     public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
+        if (keyBytes == null) {
+            return stickyPartitionCache.partition(topic, cluster);
+        } 
         List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
         int numPartitions = partitions.size();
-        if (keyBytes == null) {
-            int nextValue = nextValue(topic);
-            List<PartitionInfo> availablePartitions = cluster.availablePartitionsForTopic(topic);
-            if (!availablePartitions.isEmpty()) {
-                int part = Utils.toPositive(nextValue) % availablePartitions.size();
-                return availablePartitions.get(part).partition();
-            } else {
-                // no partitions are available, give a non-available partition
-                return Utils.toPositive(nextValue) % numPartitions;
-            }
-        } else {
-            // hash the keyBytes to choose a partition
-            return Utils.toPositive(Utils.murmur2(keyBytes)) % numPartitions;
-        }
-    }
-
-    private int nextValue(String topic) {
-        AtomicInteger counter = topicCounterMap.get(topic);
-        if (null == counter) {
-            counter = new AtomicInteger(ThreadLocalRandom.current().nextInt());
-            AtomicInteger currentCounter = topicCounterMap.putIfAbsent(topic, counter);
-            if (currentCounter != null) {
-                counter = currentCounter;
-            }
-        }
-        return counter.getAndIncrement();
+        // hash the keyBytes to choose a partition
+        return Utils.toPositive(Utils.murmur2(keyBytes)) % numPartitions;
     }
 
     public void close() {}
-
+    
+    /**
+     * Change the current sticky partition when the previous sticky partition batch is full. 
```

Comment:
This comment is a bit misleading since we might not change the current sticky partition if it is not `prevPartition`.  Also, there might not have been a previous batch.

Maybe something like "If a batch completed for the current sticky partition, change the sticky partition.  Alternately, if no sticky partition has been determined, set one."

Reply:
I think the idea I had here was that "when the previous sticky partition batch is full" is a conditional that is checked within the method. But I agree that it is not clear, so I'll change it to be more clear.